shade-test/jar-file/src/test/java/org/testcontainers/PublicAPITest.java

```diff
@@ -0,0 +1,120 @@
+package org.testcontainers;
+
+import lombok.RequiredArgsConstructor;
+import org.assertj.core.api.Assertions;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameters;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+import org.objectweb.asm.tree.ClassNode;
+import org.objectweb.asm.tree.FieldNode;
+import org.objectweb.asm.tree.MethodNode;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@RunWith(Parameterized.class)
+@RequiredArgsConstructor
+public class PublicAPITest extends AbstractJarFileTest {
```

Comment:
You won't merge your PR unless test passes, so it doesn't really matter IMO - worth trading having more verify results in single run out to get readable code IMO.

My suggestion to make the code more clear is something like this:
```
List<ClassInfo> classesInJar = getClassesFromJarFile(jarFilePath);
List<ClassInfo> publicAPIClasses = classesInJar.stream
    .filter(isPublicAccessClass)
    .filter(isTestContainerClass)
    .filter(isNotShadedTestContainerClass)
    .collect(toList());
```
Where ClassInfo is data class of: filePath, ClassNode (of asm)

Then just iterate over the list and verify for each public class the list of checks:
* superClass name doesn't isn't shaded
* method arguments aren't shaded class
etc

Anyway, that's my objective view on it.